roachpb: make the ignored seq num list copy on write

andreimatei opened the pull request:
Unfortunately the RaceTransport doesn't allow any data referenced by a
Transaction proto to be mutated. This patch makes updates to the ignored
seq nums conform.
The RaceTransport wants to prevent the server from doing such mutations.
In this case it's the client that does the mutations, which should be
allowed, but the implementation of this testing tool casts a wide net.

Release note: None

andreimatei committed:
kv, roachpb: move seq num manipulation under Transaction

It belongs better there than as a standalone function, particularly
since we need to give special consideration to transaction fields (see
next commit).

Release note: None

andreimatei committed:
roachpb: make the ignored seq num list copy on write

Unfortunately the RaceTransport doesn't allow any data referenced by a
Transaction proto to be mutated. This patch makes updates to the ignored
seq nums conform.
The RaceTransport wants to prevent the server from doing such mutations.
In this case it's the client that does the mutations, which should be
allowed, but the implementation of this testing tool casts a wide net.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45738)
<!-- Reviewable:end -->

knz reviewed (commented):
You can keep the code in the original files and keep the API, even with the new choice to copy the slice. I don't understand the motivation for the API change and file movement.

andreimatei commented:
I've moved it and changed the API for analogy with the `ObservedTimestamp` field. It's the transaction that imposes this unfortunate restriction, so I think it's good to keep all its consequences together.
Besides, there's exactly one caller to this function and it prefers the new API.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 5 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-:-M1fnBPRA3be_rK1R21X:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/roachpb/data.go, line 1201 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-M1fmRppEqLYsiOfGNz5:-M1fmRpq9WuvEOZVeq5p:b-uz0jeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e94e5553594c8dbd67b6b624b61e38ee9aee1b/pkg/roachpb/data.go#L1201)):*
> ```Go
> func (t *Transaction) AddIgnoredSeqNumRange(newRange enginepb.IgnoredSeqNumRange) {
> 	// Truncate the list at the last element not included in the new range.
> 	list := t.IgnoredSeqNums
> ```

No love for binary search? `sort.Search` makes this trivial to implement.

If there's something about the use of `AddIgnoredSeqNumRange` that makes it more likely to terminate this search at the end than at the beginning, then this is fine, but add a comment to that effect.
___
*[pkg/roachpb/data_test.go, line 1951 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-M1fn-sb41HCHgnUl_jc:-M1fn-sb41HCHgnUl_jd:b-cnv6sv) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e94e5553594c8dbd67b6b624b61e38ee9aee1b/pkg/roachpb/data_test.go#L1951)):*
> ```Go
> 		exp      []r
> 	}{
> 		{[]r{},
> ```

nit: this indentation is hard to read. Either put all three fields on the same line with the brackets or on three separate lines from the brackets.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45738)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/roachpb/data.go, line 1201 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-M1fmRppEqLYsiOfGNz5:-M1fvp2SCB0pQ7TPzjIL:b-kplcee) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e94e5553594c8dbd67b6b624b61e38ee9aee1b/pkg/roachpb/data.go#L1201)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

No love for binary search? `sort.Search` makes this trivial to implement.

If there's something about the use of `AddIgnoredSeqNumRange` that makes it more likely to terminate this search at the end than at the beginning, then this is fine, but add a comment to that effect.
</blockquote></details>

moved to `sort.Search()`.
___
*[pkg/roachpb/data_test.go, line 1951 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-M1fn-sb41HCHgnUl_jc:-M1fwD5uC9ttVy1bM6Ep:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/96e94e5553594c8dbd67b6b624b61e38ee9aee1b/pkg/roachpb/data_test.go#L1951)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: this indentation is hard to read. Either put all three fields on the same line with the brackets or on three separate lines from the brackets.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-:-M1gAahb38Iiw8wTo-fG:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-:-M1gBeFt3iQQzFlYv5x0:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz commented on pkg/roachpb/data.go:
```diff
@@ -1162,6 +1162,54 @@ func (t *Transaction) GetObservedTimestamp(nodeID NodeID) (hlc.Timestamp, bool)
+	// Truncate the list at the last element not included in the new range.
+
+	list := t.IgnoredSeqNums
+	i := sort.Search(len(list), func(i int) bool {
```
No love for sort.Search on my side. I *did* observe a bias for one side, where the binary search is inadequate. 

Please at least benchmark this to demonstrate the gain.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45738)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/roachpb/data.go, line 1203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45738#-M1cjRl4xc6AaDUUJhcc-r3-1203:-M1gNXU9B20vabOLhZ6J:b-kyrtwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8af030b3641746609f20d455abca6055f1d65aa/pkg/roachpb/data.go#L1203)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No love for sort.Search on my side. I *did* observe a bias for one side, where the binary search is inadequate. 

Please at least benchmark this to demonstrate the gain.
</blockquote></details>

I'll leave it cause the train is rolling.
I don't think the performance matters; these list will have 0 or 1 element commonly. It's slightly simpler the way it is now.


<!-- Sent from Reviewable.io -->
